// effect/dist/SchemaAST.js
/**
 * Abstract Syntax Tree (AST) representation for Effect schemas.
 *
 * This module defines the runtime data structures that represent schemas.
 * Most users work with the `Schema` module directly; use `SchemaAST` when you
 * need to inspect, traverse, or programmatically transform schema definitions.
 *
 * ## Mental model
 *
 * - **{@link AST}** — discriminated union (`_tag`) of all schema node types
 *   (e.g. `String`, `Objects`, `Union`, `Suspend`)
 * - **{@link Base}** — abstract base class shared by every node; carries
 *   annotations, checks, encoding chain, and context
 * - **{@link Encoding}** — a non-empty chain of {@link Link} values describing
 *   how to transform between the decoded (type) and encoded (wire) form
 * - **{@link Check}** — a validation filter ({@link Filter} or
 *   {@link FilterGroup}) attached to an AST node
 * - **{@link Context}** — per-property metadata: optionality, mutability,
 *   default values, key annotations
 * - **Guards** — type-narrowing predicates for each AST variant (e.g.
 *   {@link isString}, {@link isObjects})
 *
 * ## Common tasks
 *
 * - Inspect what kind of schema you have → guard functions ({@link isString},
 *   {@link isObjects}, {@link isUnion}, etc.)
 * - Get the decoded (type-level) AST → {@link toType}
 * - Get the encoded (wire-format) AST → {@link toEncoded}
 * - Swap decode/encode directions → {@link flip}
 * - Read annotations → {@link resolve}, {@link resolveAt},
 *   {@link resolveIdentifier}
 * - Build a transformation between schemas → {@link decodeTo}
 * - Add regex validation → {@link isPattern}
 *
 * ## Gotchas
 *
 * - AST nodes are structurally immutable; modification helpers return new
 *   objects via `Object.create`.
 * - {@link Arrays} represents both tuples and arrays; {@link Objects}
 *   represents both structs and records.
 * - {@link toType} and {@link toEncoded} are memoized — same input yields
 *   same output reference.
 * - {@link Suspend} lazily resolves its inner AST via a thunk; the thunk is
 *   memoized on first call.
 *
 * ## Quickstart
 *
 * **Example** (Inspecting a schema's AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.Struct({ name: Schema.String, age: Schema.Number })
 * const ast = schema.ast
 *
 * if (SchemaAST.isObjects(ast)) {
 *   console.log(ast.propertySignatures.map(ps => ps.name))
 *   // ["name", "age"]
 * }
 *
 * const encoded = SchemaAST.toEncoded(ast)
 * console.log(SchemaAST.isObjects(encoded)) // true
 * ```
 *
 * ## See also
 *
 * - {@link AST}
 * - {@link toType}
 * - {@link toEncoded}
 * - {@link flip}
 * - {@link resolve}
 *
 * @since 4.0.0
 */
import * as Arr from "./Array.js";
import * as Cause from "./Cause.js";
import * as Effect from "./Effect.js";
import * as Exit from "./Exit.js";
import { format, formatPropertyKey } from "./Formatter.js";
import { memoize } from "./Function.js";
import { effectIsExit, iterateEager } from "./internal/effect.js";
import * as internalRecord from "./internal/record.js";
import * as InternalAnnotations from "./internal/schema/annotations.js";
import * as Option from "./Option.js";
import * as Pipeable from "./Pipeable.js";
import * as Predicate from "./Predicate.js";
import * as RegEx from "./RegExp.js";
import * as Result from "./Result.js";
import * as Getter from "./SchemaGetter.js";
import * as Issue from "./SchemaIssue.js";
import * as Transformation from "./SchemaTransformation.js";
function makeGuard(tag) {
  return ast => ast._tag === tag;
}
/**
 * Returns `true` if the value is an {@link AST} node (any variant).
 *
 * Uses the internal `TypeId` brand to distinguish AST nodes from arbitrary
 * objects.
 *
 * @see {@link AST}
 *
 * @category Guard
 * @since 4.0.0
 */
export function isAST(u) {
  return Predicate.hasProperty(u, TypeId) && u[TypeId] === TypeId;
}
/**
 * Narrows an {@link AST} to {@link Declaration}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isDeclaration = /*#__PURE__*/makeGuard("Declaration");
/**
 * Narrows an {@link AST} to {@link Null}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isNull = /*#__PURE__*/makeGuard("Null");
/**
 * Narrows an {@link AST} to {@link Undefined}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isUndefined = /*#__PURE__*/makeGuard("Undefined");
/**
 * Narrows an {@link AST} to {@link Void}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isVoid = /*#__PURE__*/makeGuard("Void");
/**
 * Narrows an {@link AST} to {@link Never}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isNever = /*#__PURE__*/makeGuard("Never");
/**
 * Narrows an {@link AST} to {@link Unknown}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isUnknown = /*#__PURE__*/makeGuard("Unknown");
/**
 * Narrows an {@link AST} to {@link Any}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isAny = /*#__PURE__*/makeGuard("Any");
/**
 * Narrows an {@link AST} to {@link String}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isString = /*#__PURE__*/makeGuard("String");
/**
 * Narrows an {@link AST} to {@link Number}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isNumber = /*#__PURE__*/makeGuard("Number");
/**
 * Narrows an {@link AST} to {@link Boolean}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isBoolean = /*#__PURE__*/makeGuard("Boolean");
/**
 * Narrows an {@link AST} to {@link BigInt}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isBigInt = /*#__PURE__*/makeGuard("BigInt");
/**
 * Narrows an {@link AST} to {@link Symbol}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isSymbol = /*#__PURE__*/makeGuard("Symbol");
/**
 * Narrows an {@link AST} to {@link Literal}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isLiteral = /*#__PURE__*/makeGuard("Literal");
/**
 * Narrows an {@link AST} to {@link UniqueSymbol}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isUniqueSymbol = /*#__PURE__*/makeGuard("UniqueSymbol");
/**
 * Narrows an {@link AST} to {@link ObjectKeyword}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isObjectKeyword = /*#__PURE__*/makeGuard("ObjectKeyword");
/**
 * Narrows an {@link AST} to {@link Enum}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isEnum = /*#__PURE__*/makeGuard("Enum");
/**
 * Narrows an {@link AST} to {@link TemplateLiteral}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isTemplateLiteral = /*#__PURE__*/makeGuard("TemplateLiteral");
/**
 * Narrows an {@link AST} to {@link Arrays}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isArrays = /*#__PURE__*/makeGuard("Arrays");
/**
 * Narrows an {@link AST} to {@link Objects}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isObjects = /*#__PURE__*/makeGuard("Objects");
/**
 * Narrows an {@link AST} to {@link Union}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isUnion = /*#__PURE__*/makeGuard("Union");
/**
 * Narrows an {@link AST} to {@link Suspend}.
 *
 * @category Guard
 * @since 4.0.0
 */
export const isSuspend = /*#__PURE__*/makeGuard("Suspend");
/**
 * A single step in an {@link Encoding} chain, pairing a target {@link AST}
 * with a `Transformation` or `Middleware` that converts values between the
 * current node and the target.
 *
 * - `to` — the AST node on the other side of this transformation step.
 * - `transformation` — the bidirectional conversion logic (decode/encode).
 *
 * Links are composed into a non-empty array ({@link Encoding}) attached to
 * AST nodes that have a different encoded representation.
 *
 * @see {@link Encoding}
 * @see {@link decodeTo}
 *
 * @category model
 * @since 4.0.0
 */
export class Link {
  to;
  transformation;
  constructor(to, transformation) {
    this.to = to;
    this.transformation = transformation;
  }
}
/** @internal */
export const defaultParseOptions = {};
/**
 * Per-property metadata attached to AST nodes via {@link Base.context}.
 *
 * Tracks whether a property key is optional, mutable, has a constructor
 * default, or carries key-level annotations. Typically set by helpers like
 * {@link optionalKey} and `Schema.mutableKey`.
 *
 * - `isOptional` — the property key may be absent from the input.
 * - `isMutable` — the property is `readonly` when `false`.
 * - `defaultValue` — an {@link Encoding} applied during construction to
 *   supply missing values.
 * - `annotations` — key-level annotations (e.g. description of the key
 *   itself).
 *
 * @see {@link optionalKey}
 * @see {@link isOptional}
 *
 * @category model
 * @since 4.0.0
 */
export class Context {
  isOptional;
  isMutable;
  /** Used for constructor default values (e.g. `withConstructorDefault` API) */
  defaultValue;
  annotations;
  constructor(isOptional, isMutable, /** Used for constructor default values (e.g. `withConstructorDefault` API) */
  defaultValue = undefined, annotations = undefined) {
    this.isOptional = isOptional;
    this.isMutable = isMutable;
    this.defaultValue = defaultValue;
    this.annotations = annotations;
  }
}
const TypeId = "~effect/Schema";
/**
 * Abstract base class for all {@link AST} node variants.
 *
 * Every AST node extends `Base` and inherits these fields:
 *
 * - `annotations` — user-supplied metadata (identifier, title, description,
 *   arbitrary keys).
 * - `checks` — optional {@link Checks} for post-type-match validation.
 * - `encoding` — optional {@link Encoding} chain for type ↔ wire
 *   transformations.
 * - `context` — optional {@link Context} for per-property metadata.
 *
 * Subclasses add a `_tag` discriminant and variant-specific data.
 *
 * @see {@link AST}
 *
 * @category model
 * @since 4.0.0
 */
export class Base {
  [TypeId] = TypeId;
  annotations;
  checks;
  encoding;
  context;
  constructor(annotations = undefined, checks = undefined, encoding = undefined, context = undefined) {
    this.annotations = annotations;
    this.checks = checks;
    this.encoding = encoding;
    this.context = context;
  }
  toString() {
    return `<${this._tag}>`;
  }
}
/**
 * AST node for user-defined opaque types with custom parsing logic.
 *
 * Use when none of the built-in AST nodes fit. The `run` function receives
 * `typeParameters` and returns a parser that validates/transforms raw input.
 *
 * - `typeParameters` — inner schemas this declaration is parameterized over
 *   (e.g. the element type for a custom collection).
 * - `run` — factory producing the actual parse function.
 *
 * @see {@link isDeclaration}
 *
 * @category model
 * @since 4.0.0
 */
export class Declaration extends Base {
  _tag = "Declaration";
  typeParameters;
  run;
  constructor(typeParameters, run, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.typeParameters = typeParameters;
    this.run = run;
  }
  /** @internal */
  getParser() {
    const run = this.run(this.typeParameters);
    return (oinput, options) => {
      if (Option.isNone(oinput)) return Effect.succeedNone;
      return Effect.mapEager(run(oinput.value, this, options), Option.some);
    };
  }
  /** @internal */
  recur(recur) {
    const tps = mapOrSame(this.typeParameters, recur);
    return tps === this.typeParameters ? this : new Declaration(tps, this.run, this.annotations, this.checks, undefined, this.context);
  }
  /** @internal */
  getExpected() {
    const expected = this.annotations?.expected;
    if (typeof expected === "string") return expected;
    return "<Declaration>";
  }
}
/**
 * AST node matching the `null` literal value.
 *
 * Parsing succeeds only when the input is exactly `null`.
 *
 * @see {@link null}
 * @see {@link isNull}
 *
 * @category model
 * @since 4.0.0
 */
export class Null extends Base {
  _tag = "Null";
  /** @internal */
  getParser() {
    return fromConst(this, null);
  }
  /** @internal */
  getExpected() {
    return "null";
  }
}
const null_ = /*#__PURE__*/new Null();
export {
/**
 * Singleton {@link Null} AST instance.
 *
 * @since 4.0.0
 */
null_ as null };
/**
 * AST node matching the `undefined` value.
 *
 * Parsing succeeds only when the input is exactly `undefined`.
 *
 * @see {@link undefined}
 * @see {@link isUndefined}
 *
 * @category model
 * @since 4.0.0
 */
export class Undefined extends Base {
  _tag = "Undefined";
  /** @internal */
  getParser() {
    return fromConst(this, undefined);
  }
  /** @internal */
  toCodecJson() {
    return replaceEncoding(this, [undefinedToNull]);
  }
  /** @internal */
  getExpected() {
    return "undefined";
  }
}
const undefinedToNull = /*#__PURE__*/new Link(null_, /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.transform(() => undefined), /*#__PURE__*/Getter.transform(() => null)));
const undefined_ = /*#__PURE__*/new Undefined();
export {
/**
 * Singleton {@link Undefined} AST instance.
 *
 * @since 4.0.0
 */
undefined_ as undefined };
/**
 * AST node matching the `void` type (accepts `undefined` at runtime).
 *
 * Behaves like {@link Undefined} for parsing but represents the TypeScript
 * `void` type semantically.
 *
 * @see {@link void}
 * @see {@link isVoid}
 *
 * @category model
 * @since 4.0.0
 */
export class Void extends Base {
  _tag = "Void";
  /** @internal */
  getParser() {
    return fromConst(this, undefined);
  }
  /** @internal */
  toCodecJson() {
    return replaceEncoding(this, [undefinedToNull]);
  }
  /** @internal */
  getExpected() {
    return "void";
  }
}
const void_ = /*#__PURE__*/new Void();
export {
/**
 * Singleton {@link Void} AST instance.
 *
 * @since 4.0.0
 */
void_ as void };
/**
 * AST node representing the `never` type — no value matches.
 *
 * Parsing always fails. Useful as a placeholder in unions or as the result
 * of narrowing that eliminates all options.
 *
 * @see {@link never}
 * @see {@link isNever}
 *
 * @category model
 * @since 4.0.0
 */
export class Never extends Base {
  _tag = "Never";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isNever);
  }
  /** @internal */
  getExpected() {
    return "never";
  }
}
/**
 * Singleton {@link Never} AST instance.
 *
 * @since 4.0.0
 */
export const never = /*#__PURE__*/new Never();
/**
 * AST node representing the `any` type — every value matches.
 *
 * @see {@link any}
 * @see {@link isAny}
 *
 * @category model
 * @since 4.0.0
 */
export class Any extends Base {
  _tag = "Any";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isUnknown);
  }
  /** @internal */
  getExpected() {
    return "any";
  }
}
/**
 * Singleton {@link Any} AST instance.
 *
 * @since 4.0.0
 */
export const any = /*#__PURE__*/new Any();
/**
 * AST node representing the `unknown` type — every value matches.
 *
 * Unlike {@link Any}, this is type-safe: the parsed result is typed as
 * `unknown` rather than `any`.
 *
 * @see {@link unknown}
 * @see {@link isUnknown}
 *
 * @category model
 * @since 4.0.0
 */
export class Unknown extends Base {
  _tag = "Unknown";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isUnknown);
  }
  /** @internal */
  getExpected() {
    return "unknown";
  }
}
/**
 * Singleton {@link Unknown} AST instance.
 *
 * @since 4.0.0
 */
export const unknown = /*#__PURE__*/new Unknown();
/**
 * AST node matching the TypeScript `object` type — accepts objects, arrays,
 * and functions (anything non-primitive and non-null).
 *
 * @see {@link objectKeyword}
 * @see {@link isObjectKeyword}
 *
 * @category model
 * @since 4.0.0
 */
export class ObjectKeyword extends Base {
  _tag = "ObjectKeyword";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isObjectKeyword);
  }
  /** @internal */
  getExpected() {
    return "object | array | function";
  }
}
/**
 * Singleton {@link ObjectKeyword} AST instance.
 *
 * @since 4.0.0
 */
export const objectKeyword = /*#__PURE__*/new ObjectKeyword();
/**
 * AST node representing a TypeScript `enum`.
 *
 * Holds `enums` as an array of `[name, value]` pairs where values are
 * `string | number`. Parsing succeeds when the input matches any enum value.
 *
 * @see {@link isEnum}
 *
 * @category model
 * @since 4.0.0
 */
export class Enum extends Base {
  _tag = "Enum";
  enums;
  constructor(enums, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.enums = enums;
  }
  /** @internal */
  getParser() {
    const values = new Set(this.enums.map(([, v]) => v));
    return fromRefinement(this, input => values.has(input));
  }
  /** @internal */
  toCodecStringTree() {
    if (this.enums.some(([_, v]) => typeof v === "number")) {
      const coercions = Object.fromEntries(this.enums.map(([_, v]) => [globalThis.String(v), v]));
      return replaceEncoding(this, [new Link(new Union(Object.keys(coercions).map(k => new Literal(k)), "anyOf"), new Transformation.Transformation(Getter.transform(s => coercions[s]), Getter.String()))]);
    }
    return this;
  }
  /** @internal */
  getExpected() {
    return this.enums.map(([_, value]) => JSON.stringify(value)).join(" | ");
  }
}
function isTemplateLiteralPart(ast) {
  switch (ast._tag) {
    case "String":
    case "Number":
    case "BigInt":
    case "Literal":
    case "TemplateLiteral":
      return true;
    case "Union":
      return ast.types.every(isTemplateLiteralPart);
    default:
      return false;
  }
}
/**
 * AST node representing a TypeScript template literal type
 * (e.g. `` `user_${string}` ``).
 *
 * `parts` is an array of AST nodes; each part contributes to the
 * template literal pattern. A regex is derived from the parts to validate
 * strings at runtime.
 *
 * @see {@link isTemplateLiteral}
 *
 * @category model
 * @since 4.0.0
 */
export class TemplateLiteral extends Base {
  _tag = "TemplateLiteral";
  parts;
  /** @internal */
  encodedParts;
  constructor(parts, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    const encodedParts = [];
    for (const part of parts) {
      const encoded = toEncoded(part);
      if (isTemplateLiteralPart(encoded)) {
        encodedParts.push(encoded);
      } else {
        throw new Error(`Invalid TemplateLiteral part ${encoded._tag}`);
      }
    }
    this.parts = parts;
    this.encodedParts = encodedParts;
  }
  /** @internal */
  getParser(recur) {
    const parser = recur(this.asTemplateLiteralParser());
    return (oinput, options) => Effect.mapBothEager(parser(oinput, options), {
      onSuccess: () => oinput,
      onFailure: issue => new Issue.Composite(this, oinput, [issue])
    });
  }
  /** @internal */
  getExpected() {
    return "string";
  }
  /** @internal */
  asTemplateLiteralParser() {
    const tuple = new Arrays(false, this.parts.map(templateLiteralPartFromString), []);
    const regExp = getTemplateLiteralRegExp(this);
    return decodeTo(string, tuple, new Transformation.Transformation(Getter.transformOrFail(s => {
      const match = regExp.exec(s);
      if (match) return Effect.succeed(match.slice(1, this.parts.length + 1));
      return Effect.fail(new Issue.InvalidValue(Option.some(s), {
        message: `Expected a value matching ${regExp.source}, got ${format(s)}`
      }));
    }), Getter.transform(parts => parts.join(""))));
  }
}
/**
 * AST node matching a specific `unique symbol` value.
 *
 * Parsing succeeds only when the input is reference-equal to the stored
 * `symbol`.
 *
 * @see {@link isUniqueSymbol}
 *
 * @category model
 * @since 4.0.0
 */
export class UniqueSymbol extends Base {
  _tag = "UniqueSymbol";
  symbol;
  constructor(symbol, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.symbol = symbol;
  }
  /** @internal */
  getParser() {
    return fromConst(this, this.symbol);
  }
  /** @internal */
  toCodecStringTree() {
    return replaceEncoding(this, [symbolToString]);
  }
  /** @internal */
  getExpected() {
    return globalThis.String(this.symbol);
  }
}
/**
 * AST node matching an exact primitive value (string, number, boolean, or
 * bigint).
 *
 * Parsing succeeds only when the input is strictly equal (`===`) to the
 * stored `literal`. Numeric literals must be finite — `Infinity`, `-Infinity`,
 * and `NaN` are rejected at construction time.
 *
 * **Example** (Creating a literal AST)
 *
 * ```ts
 * import { SchemaAST } from "effect"
 *
 * const ast = new SchemaAST.Literal("active")
 * console.log(ast.literal) // "active"
 * ```
 *
 * @see {@link LiteralValue}
 * @see {@link isLiteral}
 *
 * @category model
 * @since 4.0.0
 */
export class Literal extends Base {
  _tag = "Literal";
  literal;
  constructor(literal, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    if (typeof literal === "number" && !globalThis.Number.isFinite(literal)) {
      throw new Error(`A numeric literal must be finite, got ${format(literal)}`);
    }
    this.literal = literal;
  }
  /** @internal */
  getParser() {
    return fromConst(this, this.literal);
  }
  /** @internal */
  toCodecJson() {
    return typeof this.literal === "bigint" ? literalToString(this) : this;
  }
  /** @internal */
  toCodecStringTree() {
    return typeof this.literal === "string" ? this : literalToString(this);
  }
  /** @internal */
  getExpected() {
    return typeof this.literal === "string" ? JSON.stringify(this.literal) : globalThis.String(this.literal);
  }
}
function literalToString(ast) {
  const literalAsString = globalThis.String(ast.literal);
  return replaceEncoding(ast, [new Link(new Literal(literalAsString), new Transformation.Transformation(Getter.transform(() => ast.literal), Getter.transform(() => literalAsString)))]);
}
/**
 * AST node matching any `string` value.
 *
 * @see {@link string}
 * @see {@link isString}
 *
 * @category model
 * @since 4.0.0
 */
export class String extends Base {
  _tag = "String";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isString);
  }
  /** @internal */
  getExpected() {
    return "string";
  }
}
/**
 * Singleton {@link String} AST instance.
 *
 * @since 4.0.0
 */
export const string = /*#__PURE__*/new String();
/**
 * AST node matching any `number` value (including `NaN`, `Infinity`,
 * `-Infinity`).
 *
 * Default JSON serialization:
 * - Finite numbers are serialized as JSON numbers.
 * - `Infinity`, `-Infinity`, and `NaN` are serialized as JSON strings.
 *
 * If the node has an `isFinite` or `isInt` check, the string fallback is
 * skipped since non-finite values cannot occur.
 *
 * @see {@link number}
 * @see {@link isNumber}
 *
 * @category model
 * @since 4.0.0
 */
export class Number extends Base {
  _tag = "Number";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isNumber);
  }
  /** @internal */
  toCodecJson() {
    if (this.checks && (hasCheck(this.checks, "isFinite") || hasCheck(this.checks, "isInt"))) {
      return this;
    }
    return replaceEncoding(this, [numberToJson]);
  }
  /** @internal */
  toCodecStringTree() {
    if (this.checks && (hasCheck(this.checks, "isFinite") || hasCheck(this.checks, "isInt"))) {
      return replaceEncoding(this, [finiteToString]);
    }
    return replaceEncoding(this, [numberToString]);
  }
  /** @internal */
  getExpected() {
    return "number";
  }
}
// oxlint-disable-next-line only-used-in-recursion - @gcanti what's this? :-)
function hasCheck(checks, tag) {
  return checks.some(c => {
    switch (c._tag) {
      case "Filter":
        return c.annotations?.meta?._tag === tag;
      case "FilterGroup":
        return hasCheck(c.checks, tag);
    }
  });
}
/**
 * Singleton {@link Number} AST instance.
 *
 * @since 4.0.0
 */
export const number = /*#__PURE__*/new Number();
/**
 * AST node matching any `boolean` value (`true` or `false`).
 *
 * @see {@link boolean}
 * @see {@link isBoolean}
 *
 * @category model
 * @since 4.0.0
 */
export class Boolean extends Base {
  _tag = "Boolean";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isBoolean);
  }
  /** @internal */
  getExpected() {
    return "boolean";
  }
}
/**
 * Singleton {@link Boolean} AST instance.
 *
 * @since 4.0.0
 */
export const boolean = /*#__PURE__*/new Boolean();
/**
 * AST node matching any `symbol` value.
 *
 * When serialized to a string-based codec, symbols are converted via
 * `Symbol.keyFor` and must be registered with `Symbol.for`.
 *
 * @see {@link symbol}
 * @see {@link isSymbol}
 *
 * @category model
 * @since 4.0.0
 */
export class Symbol extends Base {
  _tag = "Symbol";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isSymbol);
  }
  /** @internal */
  toCodecStringTree() {
    return replaceEncoding(this, [symbolToString]);
  }
  /** @internal */
  getExpected() {
    return "symbol";
  }
}
/**
 * Singleton {@link Symbol} AST instance.
 *
 * @since 4.0.0
 */
export const symbol = /*#__PURE__*/new Symbol();
/**
 * AST node matching any `bigint` value.
 *
 * When serialized to a string-based codec, bigints are converted to/from
 * their decimal string representation.
 *
 * @see {@link bigInt}
 * @see {@link isBigInt}
 *
 * @category model
 * @since 4.0.0
 */
export class BigInt extends Base {
  _tag = "BigInt";
  /** @internal */
  getParser() {
    return fromRefinement(this, Predicate.isBigInt);
  }
  /** @internal */
  toCodecStringTree() {
    return replaceEncoding(this, [bigIntToString]);
  }
  /** @internal */
  getExpected() {
    return "bigint";
  }
}
/**
 * Singleton {@link BigInt} AST instance.
 *
 * @since 4.0.0
 */
export const bigInt = /*#__PURE__*/new BigInt();
/**
 * AST node for array-like types — both tuples and arrays.
 *
 * - `elements` — positional element types (tuple elements). An element is
 *   optional if its {@link Context.isOptional} is `true`.
 * - `rest` — the rest/variadic element types. When non-empty, the first
 *   entry is the "spread" type (e.g. `...Array<string>`), and subsequent
 *   entries are trailing positional elements after the spread.
 * - `isMutable` — whether the resulting array is `readonly` (`false`) or
 *   mutable (`true`).
 *
 * Construction enforces TypeScript ordering rules: a required element
 * cannot follow an optional one, and an optional element cannot follow a
 * rest element.
 *
 * **Example** (Inspecting a tuple AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.Tuple([Schema.String, Schema.Number])
 * const ast = schema.ast
 *
 * if (SchemaAST.isArrays(ast)) {
 *   console.log(ast.elements.length) // 2
 *   console.log(ast.rest.length)     // 0
 * }
 * ```
 *
 * @see {@link isArrays}
 * @see {@link Objects}
 *
 * @category model
 * @since 4.0.0
 */
export class Arrays extends Base {
  _tag = "Arrays";
  isMutable;
  elements;
  rest;
  constructor(isMutable, elements, rest, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.isMutable = isMutable;
    this.elements = elements;
    this.rest = rest;
    // A required element cannot follow an optional element. ts(1257)
    const i = elements.findIndex(isOptional);
    if (i !== -1 && (elements.slice(i + 1).some(e => !isOptional(e)) || rest.length > 1)) {
      throw new Error("A required element cannot follow an optional element. ts(1257)");
    }
    // An optional element cannot follow a rest element.ts(1266)
    if (rest.length > 1 && rest.slice(1).some(isOptional)) {
      throw new Error("An optional element cannot follow a rest element. ts(1266)");
    }
  }
  /** @internal */
  getParser(recur) {
    // oxlint-disable-next-line @typescript-eslint/no-this-alias
    const ast = this;
    const elements = ast.elements.map(ast => ({
      ast,
      parser: recur(ast)
    }));
    const rest = ast.rest.map(ast => ({
      ast,
      parser: recur(ast)
    }));
    const elementLen = elements.length;
    const [head, ...tail] = rest;
    const tailLen = tail.length;
    function getParser(tailThreshold, index) {
      if (index < elementLen) {
        return elements[index];
      } else if (index >= tailThreshold) {
        return tail[index - tailThreshold];
      }
      return head;
    }
    return Effect.fnUntracedEager(function* (oinput, options) {
      if (oinput._tag === "None") {
        return oinput;
      }
      const input = oinput.value;
      // If the input is not an array, return early with an error
      if (!Array.isArray(input)) {
        return yield* Effect.fail(new Issue.InvalidType(ast, oinput));
      }
      const len = input.length;
      const state = {
        ast,
        getParser,
        oinput,
        len,
        tailThreshold: resolveTailThreshold(len, elementLen, tailLen),
        output: new globalThis.Array(len),
        issues: undefined,
        options
      };
      const concurrency = resolveConcurrency(options?.concurrency);
      const eff = parseArray(state, input, {
        concurrency: concurrency?.concurrency,
        end: ast.rest.length === 0 ? elementLen : Math.max(len, elementLen + tailLen)
      });
      if (eff) yield* eff;
      // ---------------------------------------------
      // handle excess indexes
      // ---------------------------------------------
      if (ast.rest.length === 0 && len > elementLen) {
        for (let i = elementLen; i <= len - 1; i++) {
          const issue = new Issue.Pointer([i], new Issue.UnexpectedKey(ast, input[i]));
          if (options.errors === "all") {
            if (state.issues) state.issues.push(issue);else state.issues = [issue];
          } else {
            return yield* Effect.fail(new Issue.Composite(ast, oinput, [issue]));
          }
        }
      }
      if (state.issues) {
        return yield* Effect.fail(new Issue.Composite(ast, oinput, state.issues));
      }
      return Option.some(state.output);
    });
  }
  /** @internal */
  recur(recur) {
    const elements = mapOrSame(this.elements, recur);
    const rest = mapOrSame(this.rest, recur);
    return elements === this.elements && rest === this.rest ? this : new Arrays(this.isMutable, elements, rest, this.annotations, this.checks, undefined, this.context);
  }
  /** @internal */
  getExpected() {
    return "array";
  }
}
const parseArray = /*#__PURE__*/iterateEager()({
  onItem(s, item, i) {
    const value = i < s.len ? Option.some(item) : Option.none();
    return s.getParser(s.tailThreshold, i).parser(value, s.options);
  },
  step(s, _, exit, i) {
    if (exit._tag === "Failure") {
      return wrapPropertyKeyIssue(s, s.ast, i, exit);
    } else if (exit.value._tag === "Some") {
      s.output[i] = exit.value.value;
    } else {
      const p = s.getParser(s.tailThreshold, i);
      if (isOptional(p.ast)) return;
      const issue = new Issue.Pointer([i], new Issue.MissingKey(p.ast.context?.annotations));
      if (s.options.errors === "all") {
        if (s.issues) s.issues.push(issue);else s.issues = [issue];
      } else {
        return Exit.fail(new Issue.Composite(s.ast, s.oinput, [issue]));
      }
    }
  }
});
function resolveTailThreshold(inputLen, elementLen, tailLen) {
  return Math.max(elementLen, inputLen - tailLen);
}
const resolveConcurrency = value => {
  value = value === "unbounded" ? Infinity : value ?? 1;
  return value > 1 ? {
    concurrency: value
  } : undefined;
};
const wrapPropertyKeyIssue = (s, ast, key, exit) => {
  const issueResult = Cause.findError(exit.cause);
  if (Result.isFailure(issueResult)) {
    return exit;
  }
  const issue = new Issue.Pointer([key], issueResult.success);
  if (s.options.errors === "all") {
    if (s.issues) s.issues.push(issue);else s.issues = [issue];
  } else {
    return Exit.fail(new Issue.Composite(ast, s.oinput, [issue]));
  }
};
/**
 * floating point or integer, with optional exponent
 * @internal
 */
export const FINITE_PATTERN = "[+-]?\\d*\\.?\\d+(?:[Ee][+-]?\\d+)?";
const isNumberStringRegExp = /*#__PURE__*/new globalThis.RegExp(`(?:${FINITE_PATTERN}|Infinity|-Infinity|NaN)`);
/**
 * Returns the object keys that match the index signature parameter schema.
 * @internal
 */
export function getIndexSignatureKeys(input, parameter) {
  const encoded = toEncoded(parameter);
  switch (encoded._tag) {
    case "String":
      return Object.keys(input);
    case "TemplateLiteral":
      {
        const regExp = getTemplateLiteralRegExp(encoded);
        return Object.keys(input).filter(k => regExp.test(k));
      }
    case "Symbol":
      return Object.getOwnPropertySymbols(input);
    case "Number":
      return Object.keys(input).filter(k => isNumberStringRegExp.test(k));
    case "Union":
      return [...new Set(encoded.types.flatMap(t => getIndexSignatureKeys(input, t)))];
    default:
      return [];
  }
}
/**
 * A named property within an {@link Objects} node.
 *
 * Pairs a `name` (any `PropertyKey`) with a `type` ({@link AST}). The
 * property's optionality and mutability are determined by the `type`'s
 * {@link Context}.
 *
 * @see {@link Objects}
 *
 * @category model
 * @since 4.0.0
 */
export class PropertySignature {
  name;
  type;
  constructor(name, type) {
    this.name = name;
    this.type = type;
  }
}
/**
 * Bidirectional merge strategy for index signature key-value pairs.
 *
 * Used by {@link IndexSignature} when the same key appears multiple times
 * (e.g. from `Schema.extend` or overlapping records). Provides separate
 * `decode` and `encode` combiners that determine how duplicate entries are
 * merged.
 *
 * @see {@link IndexSignature}
 *
 * @category model
 * @since 4.0.0
 */
export class KeyValueCombiner {
  decode;
  encode;
  constructor(decode, encode) {
    this.decode = decode;
    this.encode = encode;
  }
  /** @internal */
  flip() {
    return new KeyValueCombiner(this.encode, this.decode);
  }
}
/**
 * An index signature entry within an {@link Objects} node.
 *
 * - `parameter` — the key type AST (e.g. {@link String} for `string` keys,
 *   {@link TemplateLiteral} for patterned keys).
 * - `type` — the value type AST.
 * - `merge` — optional {@link KeyValueCombiner} for handling duplicate keys.
 *
 * Using `Schema.optionalKey` on the value type is not allowed for index
 * signatures (throws at construction); use `Schema.optional` instead.
 *
 * @see {@link Objects}
 * @see {@link PropertySignature}
 *
 * @category model
 * @since 4.0.0
 */
export class IndexSignature {
  parameter;
  type;
  merge;
  constructor(parameter, type, merge) {
    this.parameter = parameter;
    this.type = type;
    this.merge = merge;
    if (isOptional(type) && !containsUndefined(type)) {
      throw new Error("Cannot use `Schema.optionalKey` with index signatures, use `Schema.optional` instead.");
    }
  }
}
/**
 * AST node for object-like types — both structs and records.
 *
 * - `propertySignatures` — named properties with their types (struct fields).
 * - `indexSignatures` — index signature entries (record patterns), each with
 *   a `parameter` AST (the key type) and a `type` AST (the value type).
 *
 * An `Objects` with no properties and no index signatures acts as a bare
 * `object | array` type check (accepts any non-nullish value).
 *
 * Duplicate property names throw at construction time.
 *
 * **Example** (Inspecting a struct AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.Struct({ name: Schema.String })
 * const ast = schema.ast
 *
 * if (SchemaAST.isObjects(ast)) {
 *   for (const ps of ast.propertySignatures) {
 *     console.log(ps.name, ps.type._tag)
 *   }
 *   // "name" "String"
 * }
 * ```
 *
 * @see {@link isObjects}
 * @see {@link PropertySignature}
 * @see {@link IndexSignature}
 * @see {@link Arrays}
 *
 * @category model
 * @since 4.0.0
 */
export class Objects extends Base {
  _tag = "Objects";
  propertySignatures;
  indexSignatures;
  constructor(propertySignatures, indexSignatures, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.propertySignatures = propertySignatures;
    this.indexSignatures = indexSignatures;
    // Duplicate property signatures
    const duplicates = propertySignatures.map(ps => ps.name).filter((name, i, arr) => arr.indexOf(name) !== i);
    if (duplicates.length > 0) {
      throw new Error(`Duplicate identifiers: ${JSON.stringify(duplicates)}. ts(2300)`);
    }
  }
  /** @internal */
  getParser(recur) {
    // oxlint-disable-next-line @typescript-eslint/no-this-alias
    const ast = this;
    const expectedKeys = [];
    const expectedKeysSet = new Set();
    const properties = [];
    for (const ps of ast.propertySignatures) {
      expectedKeys.push(ps.name);
      expectedKeysSet.add(ps.name);
      properties.push({
        ps,
        parser: recur(ps.type),
        name: ps.name,
        type: ps.type
      });
    }
    const indexCount = ast.indexSignatures.length;
    // ---------------------------------------------
    // handle empty struct
    // ---------------------------------------------
    if (ast.propertySignatures.length === 0 && ast.indexSignatures.length === 0) {
      return fromRefinement(ast, Predicate.isNotNullish);
    }
    const parseIndexes = indexCount > 0 ? iterateEager()({
      onItem: Effect.fnUntracedEager(function* (s, [key, is]) {
        const parserKey = recur(indexSignatureParameterFromString(is.parameter));
        const effKey = parserKey(Option.some(key), s.options);
        const exitKey = effectIsExit(effKey) ? effKey : yield* Effect.exit(effKey);
        if (exitKey._tag === "Failure") {
          const eff = wrapPropertyKeyIssue(s, ast, key, exitKey);
          if (eff) yield* eff;
          return;
        }
        const value = Option.some(s.input[key]);
        const parserValue = recur(is.type);
        const effValue = parserValue(value, s.options);
        const exitValue = effectIsExit(effValue) ? effValue : yield* Effect.exit(effValue);
        if (exitValue._tag === "Failure") {
          const eff = wrapPropertyKeyIssue(s, ast, key, exitValue);
          if (eff) yield* eff;
          return;
        } else if (exitKey.value._tag === "Some" && exitValue.value._tag === "Some") {
          const k2 = exitKey.value.value;
          const v2 = exitValue.value.value;
          if (is.merge && is.merge.decode && Object.hasOwn(s.out, k2)) {
            const [k, v] = is.merge.decode.combine([k2, s.out[k2]], [k2, v2]);
            internalRecord.set(s.out, k, v);
          } else {
            internalRecord.set(s.out, k2, v2);
          }
        }
      }),
      step: (_s, _, exit) => exit._tag === "Failure" ? exit : undefined
    }) : undefined;
    return Effect.fnUntracedEager(function* (oinput, options) {
      if (oinput._tag === "None") {
        return oinput;
      }
      const input = oinput.value;
      // If the input is not a record, return early with an error
      if (!(typeof input === "object" && input !== null && !Array.isArray(input))) {
        return yield* Effect.fail(new Issue.InvalidType(ast, oinput));
      }
      const out = {};
      const state = {
        ast,
        oinput,
        input,
        out,
        issues: undefined,
        options
      };
      const errorsAllOption = options.errors === "all";
      const onExcessPropertyError = options.onExcessProperty === "error";
      const onExcessPropertyPreserve = options.onExcessProperty === "preserve";
      // ---------------------------------------------
      // handle excess properties
      // ---------------------------------------------
      let inputKeys;
      if (ast.indexSignatures.length === 0 && (onExcessPropertyError || onExcessPropertyPreserve)) {
        inputKeys = Reflect.ownKeys(input);
        for (let i = 0; i < inputKeys.length; i++) {
          const key = inputKeys[i];
          if (!expectedKeysSet.has(key)) {
            // key is unexpected
            if (onExcessPropertyError) {
              const issue = new Issue.Pointer([key], new Issue.UnexpectedKey(ast, input[key]));
              if (errorsAllOption) {
                if (state.issues) {
                  state.issues.push(issue);
                } else {
                  state.issues = [issue];
                }
                continue;
              } else {
                return yield* Effect.fail(new Issue.Composite(ast, oinput, [issue]));
              }
            } else {
              // preserve key
              internalRecord.set(out, key, input[key]);
            }
          }
        }
      }
      const concurrency = resolveConcurrency(options?.concurrency);
      // ---------------------------------------------
      // handle property signatures
      // ---------------------------------------------
      const eff = parseProperties(state, properties, concurrency);
      if (eff) yield* eff;
      // ---------------------------------------------
      // handle index signatures
      // ---------------------------------------------
      if (parseIndexes) {
        const keyPairs = Arr.empty();
        for (let i = 0; i < indexCount; i++) {
          const is = ast.indexSignatures[i];
          const keys = getIndexSignatureKeys(input, is.parameter);
          for (let j = 0; j < keys.length; j++) {
            const key = keys[j];
            keyPairs.push([key, is]);
          }
        }
        const eff = parseIndexes(state, keyPairs, concurrency);
        if (eff) yield* eff;
      }
      if (state.issues) {
        return yield* Effect.fail(new Issue.Composite(ast, oinput, state.issues));
      }
      if (options.propertyOrder === "original") {
        // preserve input keys order
        const keys = (inputKeys ?? Reflect.ownKeys(input)).concat(expectedKeys);
        const preserved = {};
        for (const key of keys) {
          if (Object.hasOwn(out, key)) {
            internalRecord.set(preserved, key, out[key]);
          }
        }
        return Option.some(preserved);
      }
      return Option.some(out);
    });
  }
  rebuild(recur, flipMerge) {
    const props = mapOrSame(this.propertySignatures, ps => {
      const t = recur(ps.type);
      return t === ps.type ? ps : new PropertySignature(ps.name, t);
    });
    const indexes = mapOrSame(this.indexSignatures, is => {
      const p = recur(is.parameter);
      const t = recur(is.type);
      const merge = flipMerge ? is.merge?.flip() : is.merge;
      return p === is.parameter && t === is.type && merge === is.merge ? is : new IndexSignature(p, t, merge);
    });
    return props === this.propertySignatures && indexes === this.indexSignatures ? this : new Objects(props, indexes, this.annotations, this.checks, undefined, this.context);
  }
  /** @internal */
  flip(recur) {
    return this.rebuild(recur, true);
  }
  /** @internal */
  recur(recur) {
    return this.rebuild(recur, false);
  }
  /** @internal */
  getExpected() {
    if (this.propertySignatures.length === 0 && this.indexSignatures.length === 0) return "object | array";
    return "object";
  }
}
const parseProperties = /*#__PURE__*/iterateEager()({
  onItem(s, p) {
    const value = Object.hasOwn(s.input, p.name) ? Option.some(s.input[p.name]) : Option.none();
    return p.parser(value, s.options);
  },
  step(s, p, exit) {
    if (exit._tag === "Failure") {
      return wrapPropertyKeyIssue(s, s.ast, p.name, exit);
    } else if (exit.value._tag === "Some") {
      internalRecord.set(s.out, p.name, exit.value.value);
    } else if (!isOptional(p.type)) {
      const issue = new Issue.Pointer([p.name], new Issue.MissingKey(p.type.context?.annotations));
      if (s.options.errors === "all") {
        if (s.issues) s.issues.push(issue);else s.issues = [issue];
        return;
      } else {
        return Exit.fail(new Issue.Composite(s.ast, s.oinput, [issue]));
      }
    }
  }
});
function mergeChecks(checks, b) {
  if (!checks) {
    return b.checks;
  }
  if (!b.checks) {
    return checks;
  }
  return [...checks, ...b.checks];
}
/** @internal */
export function struct(fields, checks, annotations) {
  return new Objects(Reflect.ownKeys(fields).map(key => {
    return new PropertySignature(key, fields[key].ast);
  }), [], annotations, checks);
}
/** @internal */
export function getAST(self) {
  return self.ast;
}
/** @internal */
export function tuple(elements, checks = undefined) {
  return new Arrays(false, elements.map(e => e.ast), [], undefined, checks);
}
/** @internal */
export function union(members, mode, checks) {
  return new Union(members.map(getAST), mode, undefined, checks);
}
/** @internal */
export function structWithRest(ast, records) {
  if (ast.encoding || records.some(r => r.encoding)) {
    throw new Error("StructWithRest does not support encodings");
  }
  let propertySignatures = ast.propertySignatures;
  let indexSignatures = ast.indexSignatures;
  let checks = ast.checks;
  for (const r of records) {
    propertySignatures = propertySignatures.concat(r.propertySignatures);
    indexSignatures = indexSignatures.concat(r.indexSignatures);
    checks = mergeChecks(checks, r);
  }
  return new Objects(propertySignatures, indexSignatures, undefined, checks);
}
/** @internal */
export function tupleWithRest(ast, rest) {
  if (ast.encoding) {
    throw new Error("TupleWithRest does not support encodings");
  }
  return new Arrays(ast.isMutable, ast.elements, rest, undefined, ast.checks);
}
function getCandidateTypes(ast) {
  switch (ast._tag) {
    case "Null":
      return ["null"];
    case "Undefined":
    case "Void":
      return ["undefined"];
    case "String":
    case "TemplateLiteral":
      return ["string"];
    case "Number":
      return ["number"];
    case "Boolean":
      return ["boolean"];
    case "Symbol":
    case "UniqueSymbol":
      return ["symbol"];
    case "BigInt":
      return ["bigint"];
    case "Arrays":
      return ["array"];
    case "ObjectKeyword":
      return ["object", "array", "function"];
    case "Objects":
      return ast.propertySignatures.length || ast.indexSignatures.length ? ["object"] : ["object", "array"];
    case "Enum":
      return Array.from(new Set(ast.enums.map(([, v]) => typeof v)));
    case "Literal":
      return [typeof ast.literal];
    case "Union":
      return Array.from(new Set(ast.types.flatMap(getCandidateTypes)));
    default:
      return ["null", "undefined", "string", "number", "boolean", "symbol", "bigint", "object", "array", "function"];
  }
}
/** @internal */
export function collectSentinels(ast) {
  switch (ast._tag) {
    default:
      return [];
    case "Declaration":
      {
        const s = ast.annotations?.["~sentinels"];
        return Array.isArray(s) ? s : [];
      }
    case "Objects":
      return ast.propertySignatures.flatMap(ps => {
        const type = ps.type;
        if (!isOptional(type)) {
          if (isLiteral(type)) {
            return [{
              key: ps.name,
              literal: type.literal
            }];
          }
          if (isUniqueSymbol(type)) {
            return [{
              key: ps.name,
              literal: type.symbol
            }];
          }
        }
        return [];
      });
    case "Arrays":
      return ast.elements.flatMap((e, i) => {
        return isLiteral(e) && !isOptional(e) ? [{
          key: i,
          literal: e.literal
        }] : [];
      });
    case "Suspend":
      return collectSentinels(ast.thunk());
  }
}
const candidateIndexCache = /*#__PURE__*/new WeakMap();
function getIndex(types) {
  let idx = candidateIndexCache.get(types);
  if (idx) return idx;
  idx = {};
  for (const a of types) {
    const encoded = toEncoded(a);
    if (isNever(encoded)) continue;
    const types = getCandidateTypes(encoded);
    const sentinels = collectSentinels(encoded);
    // by-type (always filled – cheap primary filter)
    idx.byType ??= {};
    for (const t of types) (idx.byType[t] ??= []).push(a);
    if (sentinels.length > 0) {
      // discriminated variants
      idx.bySentinel ??= new Map();
      for (const {
        key,
        literal
      } of sentinels) {
        let m = idx.bySentinel.get(key);
        if (!m) idx.bySentinel.set(key, m = new Map());
        let arr = m.get(literal);
        if (!arr) m.set(literal, arr = []);
        arr.push(a);
      }
    } else {
      // non-discriminated
      idx.otherwise ??= {};
      for (const t of types) (idx.otherwise[t] ??= []).push(a);
    }
  }
  candidateIndexCache.set(types, idx);
  return idx;
}
function filterLiterals(input) {
  return ast => {
    const encoded = toEncoded(ast);
    return encoded._tag === "Literal" ? encoded.literal === input : encoded._tag === "UniqueSymbol" ? encoded.symbol === input : true;
  };
}
/**
 * The goal is to reduce the number of a union members that will be checked.
 * This is useful to reduce the number of issues that will be returned.
 *
 * @internal
 */
export function getCandidates(input, types) {
  const idx = getIndex(types);
  const runtimeType = input === null ? "null" : Array.isArray(input) ? "array" : typeof input;
  // 1. Try sentinel-based dispatch (most selective)
  if (idx.bySentinel) {
    const base = idx.otherwise?.[runtimeType] ?? [];
    if (runtimeType === "object" || runtimeType === "array") {
      for (const [k, m] of idx.bySentinel) {
        if (Object.hasOwn(input, k)) {
          const match = m.get(input[k]);
          if (match) return [...match, ...base].filter(filterLiterals(input));
        }
      }
    }
    return base;
  }
  // 2. Fallback: runtime-type dispatch only
  return (idx.byType?.[runtimeType] ?? []).filter(filterLiterals(input));
}
/**
 * AST node representing a union of schemas.
 *
 * - `types` — the member AST nodes.
 * - `mode` — `"anyOf"` succeeds on the first match (like TypeScript unions);
 *   `"oneOf"` requires exactly one member to match (fails if multiple do).
 *
 * During parsing, members are tried in order. An internal candidate index
 * narrows which members to try based on the runtime type of the input and
 * discriminant ("sentinel") fields, making large unions efficient.
 *
 * **Example** (Inspecting a union AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.Union([Schema.String, Schema.Number])
 * const ast = schema.ast
 *
 * if (SchemaAST.isUnion(ast)) {
 *   console.log(ast.types.length) // 2
 *   console.log(ast.mode)         // "anyOf"
 * }
 * ```
 *
 * @see {@link isUnion}
 *
 * @category model
 * @since 4.0.0
 */
export class Union extends Base {
  _tag = "Union";
  types;
  mode;
  constructor(types, mode, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.types = types;
    this.mode = mode;
  }
  /** @internal */
  getParser(recur) {
    // oxlint-disable-next-line @typescript-eslint/no-this-alias
    const ast = this;
    return (oinput, options) => {
      if (oinput._tag === "None") {
        return Effect.succeed(oinput);
      }
      const input = oinput.value;
      const candidates = getCandidates(input, ast.types);
      const state = {
        ast,
        recur,
        oinput,
        input,
        out: undefined,
        successes: [],
        issues: undefined,
        options
      };
      const concurrency = resolveConcurrency(options?.concurrency);
      const eff = parseUnion(state, candidates, concurrency);
      if (!eff) {
        return state.out ? Effect.succeed(state.out) : Effect.fail(new Issue.AnyOf(ast, input, state.issues ?? []));
      }
      return Effect.flatMap(eff, _ => {
        return state.out ? Effect.succeed(state.out) : Effect.fail(new Issue.AnyOf(ast, input, state.issues ?? []));
      });
    };
  }
  /** @internal */
  recur(recur) {
    const types = mapOrSame(this.types, recur);
    return types === this.types ? this : new Union(types, this.mode, this.annotations, this.checks, undefined, this.context);
  }
  /** @internal */
  getExpected(getExpected) {
    const expected = this.annotations?.expected;
    if (typeof expected === "string") return expected;
    if (this.types.length === 0) return "never";
    const types = this.types.map(type => {
      const encoded = toEncoded(type);
      switch (encoded._tag) {
        case "Arrays":
          {
            const literals = encoded.elements.filter(isLiteral);
            if (literals.length > 0) {
              return `${formatIsMutable(encoded.isMutable)}[ ${literals.map(e => getExpected(e) + formatIsOptional(e.context?.isOptional)).join(", ")}, ... ]`;
            }
            break;
          }
        case "Objects":
          {
            const literals = encoded.propertySignatures.filter(ps => isLiteral(ps.type));
            if (literals.length > 0) {
              return `{ ${literals.map(ps => `${formatIsMutable(ps.type.context?.isMutable)}${formatPropertyKey(ps.name)}${formatIsOptional(ps.type.context?.isOptional)}: ${getExpected(ps.type)}`).join(", ")}, ... }`;
            }
            break;
          }
      }
      return getExpected(encoded);
    });
    return Array.from(new Set(types)).join(" | ");
  }
}
const parseUnion = /*#__PURE__*/iterateEager()({
  onItem(s, ast) {
    const parser = s.recur(ast);
    return parser(s.oinput, s.options);
  },
  step(s, candidate, exit) {
    if (exit._tag === "Failure") {
      const issueResult = Cause.findError(exit.cause);
      if (Result.isFailure(issueResult)) {
        return exit;
      }
      if (s.issues) s.issues.push(issueResult.success);else s.issues = [issueResult.success];
    } else {
      if (s.out && s.ast.mode === "oneOf") {
        s.successes.push(candidate);
        return Exit.fail(new Issue.OneOf(s.ast, s.input, s.successes));
      }
      s.out = exit.value;
      s.successes.push(candidate);
      if (s.ast.mode === "anyOf") {
        return Exit.void;
      }
    }
  }
});
const nonFiniteLiterals = /*#__PURE__*/new Union([/*#__PURE__*/new Literal("Infinity"), /*#__PURE__*/new Literal("-Infinity"), /*#__PURE__*/new Literal("NaN")], "anyOf");
const numberToJson = /*#__PURE__*/new Link(/*#__PURE__*/new Union([number, nonFiniteLiterals], "anyOf"), /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.Number(), /*#__PURE__*/Getter.transform(n => globalThis.Number.isFinite(n) ? n : globalThis.String(n))));
function formatIsMutable(isMutable) {
  return isMutable ? "" : "readonly ";
}
function formatIsOptional(isOptional) {
  return isOptional ? "?" : "";
}
/** @internal */
export function memoizeThunk(f) {
  let done = false;
  let a;
  return () => {
    if (done) {
      return a;
    }
    a = f();
    done = true;
    return a;
  };
}
/**
 * AST node for lazy/recursive schemas.
 *
 * Wraps a thunk (`() => AST`) that is memoized on first call. Use this to
 * define recursive or mutually recursive schemas without infinite loops at
 * construction time.
 *
 * **Example** (Recursive schema AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * interface Category {
 *   readonly name: string
 *   readonly children: ReadonlyArray<Category>
 * }
 *
 * const Category = Schema.Struct({
 *   name: Schema.String,
 *   children: Schema.Array(Schema.suspend((): Schema.Codec<Category> => Category))
 * })
 *
 * // The recursive branch is a Suspend node
 * ```
 *
 * @see {@link isSuspend}
 *
 * @category model
 * @since 4.0.0
 */
export class Suspend extends Base {
  _tag = "Suspend";
  thunk;
  constructor(thunk, annotations, checks, encoding, context) {
    super(annotations, checks, encoding, context);
    this.thunk = memoizeThunk(thunk);
  }
  /** @internal */
  getParser(recur) {
    return recur(this.thunk());
  }
  /** @internal */
  recur(recur) {
    return new Suspend(() => recur(this.thunk()), this.annotations, this.checks, undefined, this.context);
  }
  /** @internal */
  getExpected(getExpected) {
    return getExpected(this.thunk());
  }
}
// -----------------------------------------------------------------------------
// Checks
// -----------------------------------------------------------------------------
/**
 * A single validation check attached to an AST node.
 *
 * - `run` — the validation function. Returns `undefined` on success, or an
 *   `Issue` on failure.
 * - `annotations` — optional filter-level metadata (expected message, meta
 *   tags, arbitrary constraint hints).
 * - `aborted` — when `true`, parsing stops immediately after this filter
 *   fails (no further checks run).
 *
 * Use `.annotate()` to add metadata and `.abort()` to mark as aborting.
 * Combine with another check via `.and()` to form a {@link FilterGroup}.
 *
 * @see {@link FilterGroup}
 * @see {@link Check}
 * @see {@link isPattern}
 *
 * @category model
 * @since 4.0.0
 */
export class Filter extends Pipeable.Class {
  _tag = "Filter";
  run;
  annotations;
  /**
   * Whether the parsing process should be aborted after this check has failed.
   */
  aborted;
  constructor(run, annotations = undefined,
  /**
   * Whether the parsing process should be aborted after this check has failed.
   */
  aborted = false) {
    super();
    this.run = run;
    this.annotations = annotations;
    this.aborted = aborted;
  }
  annotate(annotations) {
    return new Filter(this.run, {
      ...this.annotations,
      ...annotations
    }, this.aborted);
  }
  abort() {
    return new Filter(this.run, this.annotations, true);
  }
  and(other, annotations) {
    return new FilterGroup([this, other], annotations);
  }
}
/**
 * A composite validation check grouping multiple {@link Check} values.
 *
 * Created by calling `.and()` on a {@link Filter} or another `FilterGroup`.
 * All inner checks are run; failures from aborted filters still stop
 * evaluation.
 *
 * @see {@link Filter}
 * @see {@link Check}
 *
 * @category model
 * @since 4.0.0
 */
export class FilterGroup extends Pipeable.Class {
  _tag = "FilterGroup";
  checks;
  annotations;
  constructor(checks, annotations = undefined) {
    super();
    this.checks = checks;
    this.annotations = annotations;
  }
  annotate(annotations) {
    return new FilterGroup(this.checks, {
      ...this.annotations,
      ...annotations
    });
  }
  and(other, annotations) {
    return new FilterGroup([this, other], annotations);
  }
}
/** @internal */
export function makeFilter(filter, annotations, aborted = false) {
  return new Filter((input, ast, options) => Issue.make(input, ast, filter(input, ast, options)), annotations, aborted);
}
/** @internal */
export function makeFilterByGuard(is, annotations) {
  return new Filter(input => is(input) ? undefined : new Issue.InvalidValue(Option.some(input)), annotations, true // after a guard, we always want to abort
  );
}
/**
 * Creates a {@link Filter} that validates strings against a regular expression.
 *
 * - Returns a `Filter<string>` suitable for use with `Schema.filter` or
 *   attached directly to a `String` AST node via checks.
 * - The regex `source` is stored in annotations for serialization and
 *   arbitrary generation.
 *
 * **Example** (Validating an email pattern)
 *
 * ```ts
 * import { SchemaAST } from "effect"
 *
 * const emailFilter = SchemaAST.isPattern(/^[^@]+@[^@]+$/)
 * ```
 *
 * @see {@link Filter}
 *
 * @since 4.0.0
 */
export function isPattern(regExp, annotations) {
  const source = regExp.source;
  return makeFilter(s => regExp.test(s), {
    expected: `a string matching the RegExp ${source}`,
    meta: {
      _tag: "isPattern",
      regExp
    },
    toArbitraryConstraint: {
      string: {
        patterns: [regExp.source]
      }
    },
    ...annotations
  });
}
function modifyOwnPropertyDescriptors(ast, f) {
  const d = Object.getOwnPropertyDescriptors(ast);
  f(d);
  return Object.create(Object.getPrototypeOf(ast), d);
}
/** @internal */
export function replaceEncoding(ast, encoding) {
  if (ast.encoding === encoding) {
    return ast;
  }
  return modifyOwnPropertyDescriptors(ast, d => {
    d.encoding.value = encoding;
  });
}
/** @internal */
export function replaceContext(ast, context) {
  if (ast.context === context) {
    return ast;
  }
  return modifyOwnPropertyDescriptors(ast, d => {
    d.context.value = context;
  });
}
/** @internal */
export function getLastEncoding(ast) {
  return ast.encoding ? getLastEncoding(ast.encoding[ast.encoding.length - 1].to) : ast;
}
/** @internal */
export function annotate(ast, annotations) {
  if (ast.checks) {
    const last = ast.checks[ast.checks.length - 1];
    return replaceChecks(ast, Arr.append(ast.checks.slice(0, -1), last.annotate(annotations)));
  }
  return modifyOwnPropertyDescriptors(ast, d => {
    d.annotations.value = {
      ...d.annotations.value,
      ...annotations
    };
  });
}
/** @internal */
export function replaceChecks(ast, checks) {
  if (ast.checks === checks) {
    return ast;
  }
  return modifyOwnPropertyDescriptors(ast, d => {
    d.checks.value = checks;
  });
}
/** @internal */
export function appendChecks(ast, checks) {
  return replaceChecks(ast, ast.checks ? [...ast.checks, ...checks] : checks);
}
function updateLastLink(encoding, f) {
  const links = encoding;
  const last = links[links.length - 1];
  const to = f(last.to);
  if (to !== last.to) {
    return Arr.append(encoding.slice(0, encoding.length - 1), new Link(to, last.transformation));
  }
  return encoding;
}
/** @internal */
export function applyToLastLink(f) {
  return ast => ast.encoding ? replaceEncoding(ast, updateLastLink(ast.encoding, f)) : ast;
}
/** @internal */
export function middlewareDecoding(ast, middleware) {
  return appendTransformation(ast, middleware, toType(ast));
}
/** @internal */
export function middlewareEncoding(ast, middleware) {
  return appendTransformation(toEncoded(ast), middleware, ast);
}
function appendTransformation(from, transformation, to) {
  const link = new Link(from, transformation);
  return replaceEncoding(to, to.encoding ? [...to.encoding, link] : [link]);
}
/** @internal */
export function brand(ast, brand) {
  const existing = InternalAnnotations.resolveBrands(ast);
  const brands = existing ? [...existing, brand] : [brand];
  return annotate(ast, {
    brands
  });
}
export function mapOrSame(as, f) {
  let changed = false;
  const out = new Array(as.length);
  for (let i = 0; i < as.length; i++) {
    const a = as[i];
    const fa = f(a);
    if (fa !== a) {
      changed = true;
    }
    out[i] = fa;
  }
  return changed ? out : as;
}
/** @internal */
export function annotateKey(ast, annotations) {
  const context = ast.context ? new Context(ast.context.isOptional, ast.context.isMutable, ast.context.defaultValue, {
    ...ast.context.annotations,
    ...annotations
  }) : new Context(false, false, undefined, annotations);
  return replaceContext(ast, context);
}
/** @internal */
export const optionalKeyLastLink = /*#__PURE__*/applyToLastLink(optionalKey);
/**
 * Marks an AST node's property key as optional by setting
 * {@link Context.isOptional} to `true`.
 *
 * Also propagates the optional flag through the last link of the encoding
 * chain if present.
 *
 * @see {@link isOptional}
 * @see {@link Context}
 *
 * @since 4.0.0
 */
export function optionalKey(ast) {
  const context = ast.context ? ast.context.isOptional === false ? new Context(true, ast.context.isMutable, ast.context.defaultValue, ast.context.annotations) : ast.context : new Context(true, false);
  return optionalKeyLastLink(replaceContext(ast, context));
}
const mutableKeyLastLink = /*#__PURE__*/applyToLastLink(mutableKey);
/** @internal */
export function mutableKey(ast) {
  const context = ast.context ? ast.context.isMutable === false ? new Context(ast.context.isOptional, true, ast.context.defaultValue, ast.context.annotations) : ast.context : new Context(false, true);
  return mutableKeyLastLink(replaceContext(ast, context));
}
/** @internal */
export function withConstructorDefault(ast, defaultValue) {
  const transformation = new Transformation.Transformation(Getter.withDefault(defaultValue), Getter.passthrough());
  const encoding = [new Link(unknown, transformation)];
  const context = ast.context ? new Context(ast.context.isOptional, ast.context.isMutable, encoding, ast.context.annotations) : new Context(false, false, encoding);
  return replaceContext(ast, context);
}
/**
 * Attaches a `Transformation` to the `to` AST, making it decode from the
 * `from` AST and encode back to it.
 *
 * This is the low-level primitive behind `Schema.transform` and
 * `Schema.transformOrFail`. It appends a {@link Link} to the `to` node's
 * encoding chain.
 *
 * - Does not mutate either input.
 * - Returns a new AST with the same type as `to`.
 *
 * @see {@link Link}
 * @see {@link Encoding}
 * @see {@link flip}
 *
 * @since 4.0.0
 */
export function decodeTo(from, to, transformation) {
  return appendTransformation(from, transformation, to);
}
function parseParameter(ast) {
  switch (ast._tag) {
    case "Literal":
      return {
        literals: Predicate.isPropertyKey(ast.literal) ? [ast.literal] : [],
        parameters: []
      };
    case "UniqueSymbol":
      return {
        literals: [ast.symbol],
        parameters: []
      };
    case "String":
    case "Number":
    case "Symbol":
    case "TemplateLiteral":
      return {
        literals: [],
        parameters: [ast]
      };
    case "Union":
      {
        const out = {
          literals: [],
          parameters: []
        };
        for (let i = 0; i < ast.types.length; i++) {
          const parsed = parseParameter(ast.types[i]);
          out.literals = out.literals.concat(parsed.literals);
          out.parameters = out.parameters.concat(parsed.parameters);
        }
        return out;
      }
  }
  return {
    literals: [],
    parameters: []
  };
}
/** @internal */
export function record(key, value, keyValueCombiner) {
  const {
    literals,
    parameters: indexSignatures
  } = parseParameter(key);
  return new Objects(literals.map(literal => new PropertySignature(literal, value)), indexSignatures.map(parameter => new IndexSignature(parameter, value, keyValueCombiner)));
}
// -------------------------------------------------------------------------------------
// Public APIs
// -------------------------------------------------------------------------------------
/**
 * Returns `true` if the AST node represents an optional property.
 *
 * Checks `ast.context?.isOptional`. Defaults to `false` when no
 * {@link Context} is set.
 *
 * @see {@link optionalKey}
 * @see {@link Context}
 *
 * @since 4.0.0
 */
export function isOptional(ast) {
  return ast.context?.isOptional ?? false;
}
/** @internal */
export function isMutable(ast) {
  return ast.context?.isMutable ?? false;
}
/**
 * Strips all encoding transformations from an AST, returning the decoded
 * (type-level) representation.
 *
 * - Memoized: same input reference → same output reference.
 * - Recursively walks into composite nodes ({@link Arrays}, {@link Objects},
 *   {@link Union}, {@link Suspend}).
 * - Does not mutate the input.
 *
 * **Example** (Getting the type AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.NumberFromString
 * const typeAst = SchemaAST.toType(schema.ast)
 * console.log(typeAst._tag) // "Number"
 * ```
 *
 * @see {@link toEncoded}
 * @see {@link flip}
 *
 * @since 4.0.0
 */
export const toType = /*#__PURE__*/memoize(ast => {
  if (ast.encoding) {
    return toType(replaceEncoding(ast, undefined));
  }
  const out = ast;
  return out.recur?.(toType) ?? out;
});
/**
 * Returns the encoded (wire-format) AST by flipping and then stripping
 * encodings.
 *
 * Equivalent to `toType(flip(ast))`. This gives you the AST that describes
 * the shape of the serialized/encoded data.
 *
 * - Memoized: same input reference → same output reference.
 * - Does not mutate the input.
 *
 * **Example** (Getting the encoded AST)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.NumberFromString
 * const encodedAst = SchemaAST.toEncoded(schema.ast)
 * console.log(encodedAst._tag) // "String"
 * ```
 *
 * @see {@link toType}
 * @see {@link flip}
 *
 * @since 4.0.0
 */
export const toEncoded = /*#__PURE__*/memoize(ast => {
  return toType(flip(ast));
});
function flipEncoding(ast, encoding) {
  const links = encoding;
  const len = links.length;
  const last = links[len - 1];
  const ls = [new Link(flip(replaceEncoding(ast, undefined)), links[0].transformation.flip())];
  for (let i = 1; i < len; i++) {
    ls.unshift(new Link(flip(links[i - 1].to), links[i].transformation.flip()));
  }
  const to = flip(last.to);
  if (to.encoding) {
    return replaceEncoding(to, [...to.encoding, ...ls]);
  } else {
    return replaceEncoding(to, ls);
  }
}
/**
 * Swaps the decode and encode directions of an AST's {@link Encoding} chain.
 *
 * After flipping, what was decoding becomes encoding and vice versa. This is
 * the core operation behind `Schema.encode` — encoding a value is decoding
 * with a flipped AST.
 *
 * - Memoized: same input reference → same output reference.
 * - Recursively walks composite nodes.
 * - Does not mutate the input.
 *
 * @see {@link toType}
 * @see {@link toEncoded}
 *
 * @since 4.0.0
 */
export const flip = /*#__PURE__*/memoize(ast => {
  if (ast.encoding) {
    return flipEncoding(ast, ast.encoding);
  }
  const out = ast;
  return out.flip?.(flip) ?? out.recur?.(flip) ?? out;
});
/** @internal */
export function containsUndefined(ast) {
  switch (ast._tag) {
    case "Undefined":
      return true;
    case "Union":
      return ast.types.some(containsUndefined);
    default:
      return false;
  }
}
function getTemplateLiteralSource(ast, top) {
  return ast.encodedParts.map(part => handleTemplateLiteralASTPartParens(part, getTemplateLiteralASTPartPattern(part), top)).join("");
}
/** @internal */
export const getTemplateLiteralRegExp = /*#__PURE__*/memoize(ast => {
  return new globalThis.RegExp(`^${getTemplateLiteralSource(ast, true)}$`);
});
function getTemplateLiteralASTPartPattern(part) {
  switch (part._tag) {
    case "Literal":
      return RegEx.escape(globalThis.String(part.literal));
    case "String":
      return STRING_PATTERN;
    case "Number":
      return FINITE_PATTERN;
    case "BigInt":
      return BIGINT_PATTERN;
    case "TemplateLiteral":
      return getTemplateLiteralSource(part, false);
    case "Union":
      return part.types.map(getTemplateLiteralASTPartPattern).join("|");
  }
}
function handleTemplateLiteralASTPartParens(part, s, top) {
  if (isUnion(part)) {
    if (!top) {
      return `(?:${s})`;
    }
  } else if (!top) {
    return s;
  }
  return `(${s})`;
}
function fromConst(ast, value) {
  const succeed = Effect.succeedSome(value);
  return oinput => {
    if (oinput._tag === "None") {
      return Effect.succeedNone;
    }
    return oinput.value === value ? succeed : Effect.fail(new Issue.InvalidType(ast, oinput));
  };
}
function fromRefinement(ast, refinement) {
  return oinput => {
    if (oinput._tag === "None") {
      return Effect.succeedNone;
    }
    return refinement(oinput.value) ? Effect.succeed(oinput) : Effect.fail(new Issue.InvalidType(ast, oinput));
  };
}
/** @internal */
export const enumsToLiterals = /*#__PURE__*/memoize(ast => {
  return new Union(ast.enums.map(e => new Literal(e[1], {
    title: e[0]
  })), "anyOf");
});
/** @internal */
export function toCodec(f) {
  function out(ast) {
    return ast.encoding ? replaceEncoding(ast, updateLastLink(ast.encoding, out)) : f(ast);
  }
  return memoize(out);
}
const indexSignatureParameterFromString = /*#__PURE__*/toCodec(ast => {
  switch (ast._tag) {
    default:
      return ast;
    case "Number":
      return ast.toCodecStringTree();
    case "Union":
      return ast.recur(indexSignatureParameterFromString);
  }
});
const templateLiteralPartFromString = /*#__PURE__*/toCodec(ast => {
  switch (ast._tag) {
    default:
      return ast;
    case "String":
    case "TemplateLiteral":
      return ast;
    case "BigInt":
    case "Number":
    case "Literal":
      return ast.toCodecStringTree();
    case "Union":
      return ast.recur(templateLiteralPartFromString);
  }
});
/**
 * any string, including newlines
 * @internal
 */
export const STRING_PATTERN = "[\\s\\S]*?";
const isStringFiniteRegExp = /*#__PURE__*/new globalThis.RegExp(`^${FINITE_PATTERN}$`);
/** @internal */
export function isStringFinite(annotations) {
  return isPattern(isStringFiniteRegExp, {
    expected: "a string representing a finite number",
    meta: {
      _tag: "isStringFinite",
      regExp: isStringFiniteRegExp
    },
    ...annotations
  });
}
const finiteString = /*#__PURE__*/appendChecks(string, [/*#__PURE__*/isStringFinite()]);
const finiteToString = /*#__PURE__*/new Link(finiteString, Transformation.numberFromString);
const numberToString = /*#__PURE__*/new Link(/*#__PURE__*/new Union([finiteString, nonFiniteLiterals], "anyOf"), Transformation.numberFromString);
/**
 * signed integer only (no leading "+" because TypeScript doesn't support it)
 */
const BIGINT_PATTERN = "-?\\d+";
const isStringBigIntRegExp = /*#__PURE__*/new globalThis.RegExp(`^${BIGINT_PATTERN}$`);
/** @internal */
export function isStringBigInt(annotations) {
  return isPattern(isStringBigIntRegExp, {
    expected: "a string representing a bigint",
    meta: {
      _tag: "isStringBigInt",
      regExp: isStringBigIntRegExp
    },
    ...annotations
  });
}
/** @internal */
export const bigIntString = /*#__PURE__*/appendChecks(string, [/*#__PURE__*/isStringBigInt({
  expected: "a string representing a bigint"
})]);
const bigIntToString = /*#__PURE__*/new Link(bigIntString, Transformation.bigintFromString);
const REGEXP_PATTERN = "Symbol\\((.*)\\)";
const isStringSymbolRegExp = /*#__PURE__*/new globalThis.RegExp(`^${REGEXP_PATTERN}$`);
/** @internal */
export const symbolString = /*#__PURE__*/appendChecks(string, [/*#__PURE__*/isStringSymbol()]);
/**
 * to distinguish between Symbol and String, we need to add a check to the string keyword
 */
const symbolToString = /*#__PURE__*/new Link(symbolString, /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.transform(description => globalThis.Symbol.for(isStringSymbolRegExp.exec(description)[1])), /*#__PURE__*/Getter.transformOrFail(sym => {
  const key = globalThis.Symbol.keyFor(sym);
  if (key !== undefined) {
    return Effect.succeed(globalThis.String(sym));
  }
  return Effect.fail(new Issue.Forbidden(Option.some(sym), {
    message: "cannot serialize to string, Symbol is not registered"
  }));
})));
/** @internal */
export function isStringSymbol(annotations) {
  return isPattern(isStringSymbolRegExp, {
    expected: "a string representing a symbol",
    meta: {
      _tag: "isStringSymbol",
      regExp: isStringSymbolRegExp
    },
    ...annotations
  });
}
/** @internal */
export function collectIssues(checks, value, issues, ast, options) {
  for (let i = 0; i < checks.length; i++) {
    const check = checks[i];
    if (check._tag === "FilterGroup") {
      collectIssues(check.checks, value, issues, ast, options);
    } else {
      const issue = check.run(value, ast, options);
      if (issue) {
        issues.push(new Issue.Filter(value, check, issue));
        if (check.aborted || options?.errors !== "all") {
          return;
        }
      }
    }
  }
}
/** @internal */
export function runChecks(checks, s) {
  const issues = [];
  collectIssues(checks, s, issues, unknown, {
    errors: "all"
  });
  if (Arr.isArrayNonEmpty(issues)) {
    const issue = new Issue.Composite(unknown, Option.some(s), issues);
    return Result.fail(issue);
  }
  return Result.succeed(s);
}
/** @internal */
export const ClassTypeId = "~effect/Schema/Class";
/** @internal */
export const STRUCTURAL_ANNOTATION_KEY = "~structural";
/**
 * Returns all annotations from the AST node.
 *
 * If the node has {@link Checks}, returns annotations from the last check
 * (which is where user-supplied annotations end up after `.pipe(Schema.annotations(...))`).
 * Otherwise returns `Base.annotations` directly.
 *
 * **Example** (Reading annotations)
 *
 * ```ts
 * import { Schema, SchemaAST } from "effect"
 *
 * const schema = Schema.String.annotate({ title: "Name" })
 * const annotations = SchemaAST.resolve(schema.ast)
 * console.log(annotations?.title) // "Name"
 * ```
 *
 * @see {@link resolveAt}
 * @see {@link resolveIdentifier}
 * @see {@link resolveTitle}
 * @see {@link resolveDescription}
 *
 * @since 4.0.0
 */
export const resolve = InternalAnnotations.resolve;
/**
 * Returns a single annotation value by key from the AST node.
 *
 * Like {@link resolve}, reads from the last check's annotations when checks
 * are present. Returns `undefined` if the key is not found.
 *
 * @see {@link resolve}
 *
 * @since 4.0.0
 */
export const resolveAt = InternalAnnotations.resolveAt;
/**
 * Returns the `identifier` annotation from the AST node, if set.
 *
 * The identifier is typically set by `Schema.annotations({ identifier: "..." })`
 * and is used for error messages and schema identification.
 *
 * @see {@link resolve}
 * @see {@link resolveTitle}
 *
 * @since 4.0.0
 */
export const resolveIdentifier = InternalAnnotations.resolveIdentifier;
/**
 * Returns the `title` annotation from the AST node, if set.
 *
 * @see {@link resolve}
 * @see {@link resolveIdentifier}
 * @see {@link resolveDescription}
 *
 * @since 4.0.0
 */
export const resolveTitle = InternalAnnotations.resolveTitle;
/**
 * Returns the `description` annotation from the AST node, if set.
 *
 * @see {@link resolve}
 * @see {@link resolveTitle}
 * @see {@link resolveIdentifier}
 *
 * @since 4.0.0
 */
export const resolveDescription = InternalAnnotations.resolveDescription;
/**
 * Returns true if the value is a JSON value.
 *
 * When a cyclic reference is detected, returns false.
 *
 * @internal
 */
export function isJson(u) {
  // `onPath` is the current recursion stack: nodes between the root and the
  // one being visited. A hit here means we looped back to an ancestor — a
  // real cycle, not a DAG — so the value is not JSON.
  const onPath = new Set();
  // `validated` memoizes subtrees we've already fully checked. Without it, a
  // diamond-shaped DAG (same node reached through multiple parents) would be
  // re-traversed once per parent, which is exponential in the nesting depth.
  const validated = new Set();
  return recur(u);
  function recur(u) {
    if (u === null || typeof u === "string" || typeof u === "boolean") {
      return true;
    }
    if (typeof u === "number") {
      return globalThis.Number.isFinite(u);
    }
    if (typeof u !== "object" || u === undefined) {
      return false;
    }
    if (onPath.has(u)) {
      return false;
    }
    if (validated.has(u)) {
      return true;
    }
    onPath.add(u);
    const ok = Array.isArray(u) ? u.every(recur) : Object.keys(u).every(key => recur(u[key]));
    // Pop on exit so siblings reaching the same node via a different path
    // don't see it as an ancestor (that would reject valid DAGs).
    onPath.delete(u);
    if (ok) {
      validated.add(u);
    }
    return ok;
  }
}
/** @internal */
export const Json = /*#__PURE__*/new Declaration([], () => (input, ast) => isJson(input) ? Effect.succeed(input) : Effect.fail(new Issue.InvalidType(ast, Option.some(input))), {
  typeConstructor: {
    _tag: "effect/Json"
  },
  generation: {
    runtime: `Schema.Json`,
    Type: `Schema.Json`
  },
  expected: "JSON value",
  toCodecJson: () => new Link(unknown, Transformation.passthrough())
});
/** @internal */
export const MutableJson = /*#__PURE__*/annotate(Json, {
  typeConstructor: {
    _tag: "effect/MutableJson"
  },
  generation: {
    runtime: `Schema.MutableJson`,
    Type: `Schema.MutableJson`
  }
});
/** @internal */
export const unknownToNull = /*#__PURE__*/new Link(null_, /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.passthrough(), /*#__PURE__*/Getter.transform(() => null)));
/** @internal */
export const unknownToJson = /*#__PURE__*/new Link(Json, /*#__PURE__*/Transformation.passthrough());
/**
 * Returns true if the value is a StringTree value.
 *
 * When a cyclic reference is detected, returns false.
 *
 * @internal
 */
export function isStringTree(u) {
  const seen = new Set();
  return recur(u);
  function recur(u) {
    if (u === undefined || typeof u === "string") {
      return true;
    }
    if (typeof u !== "object" || u === null) {
      return false;
    }
    if (seen.has(u)) {
      return false;
    }
    seen.add(u);
    if (Array.isArray(u)) {
      return u.every(recur);
    }
    return Object.keys(u).every(key => recur(u[key]));
  }
}
const StringTree = /*#__PURE__*/new Declaration([], () => (input, ast) => isStringTree(input) ? Effect.succeed(input) : Effect.fail(new Issue.InvalidType(ast, Option.some(input))), {
  expected: "StringTree",
  toCodecStringTree: () => new Link(unknown, Transformation.passthrough())
});
/** @internal */
export const unknownToStringTree = /*#__PURE__*/new Link(StringTree, /*#__PURE__*/Transformation.passthrough());
